plutus-conformance/agda/Spec.hs

```diff
@@ -3,91 +3,92 @@
-import PlutusCore.Default
-import PlutusCore.Evaluation.Machine.CostModelInterface
-import PlutusCore.Evaluation.Machine.ExBudget
-import PlutusCore.Evaluation.Machine.ExBudgetingDefaults (defaultCekMachineCosts)
-import PlutusCore.Evaluation.Machine.ExMemory
-import PlutusCore.Evaluation.Machine.SimpleBuiltinCostModel
+import PlutusCore.Default (DefaultFun, DefaultUni)
+import PlutusCore.Evaluation.Machine.CostModelInterface (CekMachineCosts, CostModelParams,
+                                                         applyCostModelParams)
+import PlutusCore.Evaluation.Machine.ExBudget (ExBudget (..))
+import PlutusCore.Evaluation.Machine.ExBudgetingDefaults (defaultCekCostModel)
+import PlutusCore.Evaluation.Machine.ExMemory (ExCPU (..), ExMemory (..))
+import PlutusCore.Evaluation.Machine.MachineParameters (CostModel (..))
+import PlutusCore.Evaluation.Machine.SimpleBuiltinCostModel (BuiltinCostKeyMap, BuiltinCostMap,
+                                                             toSimpleBuiltinCostModel)
 
 import PlutusCore.Quote
 import UntypedPlutusCore qualified as UPLC
 import UntypedPlutusCore.DeBruijn
 
--- -- | Our `evaluator` for the Agda UPLC tests is the CEK machine without costs.
--- agdaEvalUplcProgNoCost :: UplcEvaluator
--- agdaEvalUplcProgNoCost = UplcEvaluatorWithoutCosting $ \(UPLC.Program () version tmU) ->
---     let
---         -- turn it into an untyped de Bruijn term
---         tmUDB :: ExceptT FreeVariableError Quote (UPLC.Term NamedDeBruijn DefaultUni DefaultFun ())
---         tmUDB = deBruijnTerm tmU
---     in
---     case runQuote $ runExceptT $ withExceptT FreeVariableErrorE tmUDB of
---         -- if there's an exception, evaluation failed, should return `Nothing`.
---         Left _ -> Nothing
---         -- evaluate the untyped term with CEK
---         Right tmUDBSuccess ->
---             case runUAgda tmUDBSuccess of
---                 Left _ -> Nothing
---                 Right tmEvaluated ->
---                     let tmNamed = runQuote $ runExceptT $
---                             withExceptT FreeVariableErrorE $ unDeBruijnTerm tmEvaluated
---                     in
---                     -- turn it back into a named term
---                     case tmNamed of
---                         Left _          -> Nothing
---                         Right namedTerm -> Just $ UPLC.Program () version namedTerm
+import Data.Aeson (Result (Error, Success), fromJSON, toJSON)
 
--- | Our `evaluator` for the Agda UPLC tests is the CEK machine with costs.
-agdaEvalUplcProg :: UplcEvaluator
-agdaEvalUplcProg = UplcEvaluatorWithCosting $ \ modelParams (UPLC.Program () version tmU) ->
-    let
-        -- turn it into an untyped de Bruijn term
-        tmUDB :: ExceptT FreeVariableError Quote (UPLC.Term NamedDeBruijn DefaultUni DefaultFun ())
-        tmUDB = deBruijnTerm tmU
+-- This type corresponds to Cost.Raw.RawCostModel from the Agda code.
+type RawCostModel = (CekMachineCosts, BuiltinCostMap)
```

Comment:
I don't know if there's a better way to link this up with the corresponding type from Agda.  I tried experimenting with `COMPILE GHC` but didn't get anywhere.

Reply:
if you are referring to the type synonym, you could import it from the `Opts.hs` file in the metatheory.

Reply:
I'm still not sure about this.  The type `RawCostModel = (CekMachineCosts, BuiltinCostMap)` here corresponds to `RawCostModel = HCekMachineCosts × BuiltinCostMap` in the Agda code and I was wondering if there was some more "correct" way to say that they're really the same.  However, it works as it is, so maybe that's not necessary.